pandas/core/arrays/base.py

```diff
@@ -610,3 +618,121 @@ def _ndarray_values(self):
         used for interacting with our indexers.
         """
         return np.array(self)
+
+
+class ExtensionOpsBase(object):
+    """
+    A base class for the mixins for different operators.
+    Can also be used to define an individual method for a specific
+    operator using the class method create_method()
+    """
+    @classmethod
+    def create_method(cls, op):
+        """
+        A class method that returns a method that will correspond to an
+        operator for an ExtensionArray subclass.
+
+        Parameters
+        ----------
+        op:   An operator that takes arguments op(a, b)
+
+        Returns
+        -------
+        A method that can be bound to a method of a class
+
+        Usage
+        -----
+        Given an ExtensionArray subclass called MyClass, use
+
+        mymethod = create_method(my_operator)
+        in the class definition of MyClass to create the operator
+
+        """
+        op_name = ops._get_op_name(op, False)
+
+        def _binop(self, other):
+            def convert_values(parm):
+                if isinstance(parm, ExtensionArray):
+                    ovalues = list(parm)
+                elif is_extension_array_dtype(parm):
+                    ovalues = parm.values
+                elif is_list_like(parm):
+                    ovalues = parm
+                else:  # Assume its an object
+                    ovalues = [parm] * len(self)
+                return ovalues
+            lvalues = convert_values(self)
+            rvalues = convert_values(other)
+
+            try:
+                res = [op(a, b) for (a, b) in zip(lvalues, rvalues)]
+            except TypeError:
+                msg = ("ExtensionDtype invalid operation " +
+                       "{opn} between {one} and {two}")
+                raise TypeError(msg.format(opn=op_name,
+                                           one=type(lvalues),
+                                           two=type(rvalues)))
+
+            res_values = com._values_from_object(res)
+
+            try:
+                res_values = self._from_sequence(res_values)
+            except TypeError:
+                pass
+
+            return res_values
+
+        name = '__{name}__'.format(name=op_name)
```

Comment:
why are you defining create_method as anything other than NotImplementedError

this is way too opionated - if u want to define a function to do this and an author can use this ok
but it should not be the default

maybe if u want to create an ExtensionOpsMixinForScalars (long name though)

Reply:
Please take a closer look at the diff if you give such an opinionated (pun intended :-)) comment

- this is not the default, it's an optional mixin extension authors can reuse (so what you suggest?)
- so it is exactly the point to be opinionated and fallback to the scalars (we might try to think of a better name that reflects this, as you suggested)
- it is what we discussed in the other PR

Reply:
@jorisvandenbossche maybe I wasn't clear.

``ExtensionOpsBase`` should simply have NotImplementedError for the actual implementation of the arithmethic and comaprison ops. Scalar implementations of this will IMHO rarely be used and simply cluttering things up. This can be put as a separate MixIn class if you wish for inclusion by an author, including it here is forcing a pretty bad idiom: coupling the interface and implementation, and is quite error prone because its not opt in. If an author extends and adds ops, they immediate get an implementation, which could very easiy be buggy / not work / work in unexpected ways.

I think the ops implementation is so important that this must be an explicit choice. 

The point of the mixin is to avoid the boilerplate of writing ``__div__ - ....`` and so on. NOT the implementation.

Reply:
> This can be put as a separate MixIn class if you wish for inclusion by an author, 

Again, this is what this PR is doing. 
To repeat from above, is might be it is just the name of ExtensionOpsBase which is not clear? (with which I agree)

Reply:
@jreback @jorisvandenbossche Maybe the naming is misleading.  So let me clarify the intent of the PR.

The goal here is that if someone extends `ExtensionArray` (let's call it `MyArray` for sake of discussion) where they are storing objects (let's call them `MyClass`) that already have the arithmetic and/or comparison operators defined for the class `MyClass`, they can use these mixins to get those operators defined on the `MyArray` subclass by simply adding the mixin as one of the base classes of `MyArray`.

An example is `Decimal`.  The operators are defined there, so the mixin gives you the operators on `DecimalArray` "for free".  I also have my own application that extends `ExtensionArray` and it uses the same principal.

The use of these 3 classes (`ExtensionOpsBase`, `ExtensionArithmeticMixin` and `ExtensionComparisonMixin`) is *optional* for people who are writing their own subclass of `ExtensionArray`.  The `ExtensionOpsBase` class is just a base class for the 2 mixin classes.  It contains the `create_method()` method that allows one to create an individual operator, under the assumption that `MyClass` has the operator defined.  Aside from the use of the two mixin classes that define all of the operators, `ExtensionOpsBase.createmethod()` can be used to define *individual* methods.  So, for example, if the `MyClass` class only supports addition, you can choose to not use the mixin, and just do something like the following in `MyArray` to create the operator.
```python
    __add__ = ExtensionOpsBase.create_method(operator.add)
```

IMHO, it seems that @jorisvandenbossche understands what I'm trying to do, but @jreback may not fully understand it.

If you think there is a better name for `ExtensionOpsBase`, I'm open to suggestions.

Reply:
To add, @jreback, if you don't like the exact implemenation with the three classes as it is now (and not the *idea* that @Dr-Irv explained above), we can certainly discuss that, no problem at all. 

I am also thinking if it couldn't be simplified. The ExtensionOpsBase is only there (or at least as a public class) in case you want to define *some* of them and not all, and thus cannot inherit the full ExtensionArithmeticMixin class. 
This is maybe an even more specific use case, so we can maybe leave that out for now. Although implementation wise, it's probably still the cleanest to have this class as a base class for the two mixin classes.

So for naming, we can maybe go for

```
ExtensionScalarOpsMixin
ExtensionScalarArithmeticMixin
ExtensionScalarComparisonMixin
```

? 

(so adding "scalar" in there, to indicate it uses the scalar values' ops, and rename the first one from Base to Mixin)

Reply:
@jorisvandenbossche I've tried a variety of implementations without the base class, and couldn't get it to work.  I'll rename as you suggest.

Reply:
renaming looks ok to me

Reply:
Agreed on the naming conventions.  Maybe the NotImplemented versions of the methods belong on the base `ExtensionArray`?

Reply:
@jbrockmendel The NotImplemented versions are there by default for any class, so I don't think we need to do anything.  In other words, if you write a class `MyClass` that extends `ExtensionArray` and you don't define the methods for operators, they are by default NotImplemented.